### DISCUSSION TITLE
Memory tests for 8086 era computers

### DISCUSSION THREAD

**Discussion by Vutshi** [jbruchon/elks#1465] [General]
Hello everyone,

I have a question which is only tangentially related to ELKS. I try to run ELKS on an ancient PC which is partially falling apart by now. Occasionally I get an error from the BIOS memory test as some memory chips are probably not very healthy.
Hence I need a program to thoroughly test the RAM and, ideally, to identify the bad chips. It appears to be not so easy to find such a program for 8088 CPU with only 256K RAM.
Is anybody aware of any analogs of memtest86+ for 16 bit hardware?

**Comment by Mellvik** [comment]
Hi Denis,
not an unusual situation for an archeologist …

There are zillions of memory test programs for dos out there. Presuming you can actually boot DOS, Checkit3.0 is a good choice (and my choice). It does a lot of hardware testing besides memory tests. What I don't know is whether it will fit on a 360k floppy - if that's your limitation.

—M

**Reply by Vutshi** [reply]
Thank you @Mellvik,
I checked CheckIt, so to speak. It works. Although versions 3.0 and 2.0 are made for very fancy PCs with more than 256K RAM. 
In the end of the day, CheckIt 1.10A works OK but not perfect. It seems to be not able to test first two blocks of memory where the program resides.
![IMG_1551209CE6F0-1](https://user-images.githubusercontent.com/4971779/204900265-74adab94-29d4-4d5c-b238-f0b1b83e0a6b.jpeg)

Also there is an interesting option to find Bad RAM chips but it doesn't work for me neither in real hardware nor in emulation.

Somehow an open source memory test for 16 bit systems did not managed to emerge yet.

**Comment by Mellvik** [comment]
Hi @vutshi,Congratulations! Interesting results indeed.Given the errors you're finding in blocks 1 and 2, it's not surprising that some functions in checkit fail. It's more like surprising that it runs at all...I'm not a memory expert, but find it interesting that the errors occur at the same address in 2 different blocks. Given the age of your system I would look for shorts on the pcb and check the capacitors likely positioned near the end of each ram chip.Do you have a picture of the board?-M30. nov. 2022 kl. 21:27 skrev Denis Vasilyev ***@***.***>:﻿
Thank you @Mellvik,
I checked CheckIt, so to speak. It works. Although versions 3.0 and 2.0 are made for very fancy PCs with more than 256K RAM.
In the end of the day, CheckIt 1.10A works OK but not perfect. It seems to be not able to test first two blocks of memory where the program resides.

Also there is an interesting option to find Bad RAM chips but it doesn't work for me neither in real hardware nor in emulation.
Somehow an open source memory test for 16 bit systems did not managed to emerge yet.

**Reply by Vutshi** [reply]
@Mellvik,
I am not sure about these two errors. The addresses of the errors depend on the OS I am using. However, they always come in pairs like in the screenshot. Furthermore, it seems CheckIt does not consider these errors as real errors. In the message it says that they can be probably disregarded and in the very end it reports no errors. Maybe it has something to do with the fact that CHECKIT program can not be relocated...

**Comment by toncho11** [comment]
Hi @Vutshi,

I have experience testing RAM in DOS, but I will recommend using a specialized device (how it sounds, but it is simply an Arduino).
Here is an example: https://www.ebay.fr/itm/124803897060
You chips should be of type 4164 and 41256.

You can learn more here: https://www.minuszerodegrees.net/5160/ram/5160_ram_flaw_use_of_checkit.htm
There is both the Checkit method and with the debug program. Yours is not 5160, so the memory banks on the motherboard will not be the same.

Actually a memory test program for ELKS will be useful.

**Comment by Mellvik** [comment]
Thank you, @toncho11, very interesting and useful references.And yes, a memory test program for ELKS would be welcome. I'm sure there is something decent out there that can be ported.-M7. des. 2022 kl. 00:15 skrev toncho11 ***@***.***>:﻿
Hi @Vutshi,
I have experience testing RAM in DOS, but I will recommend using a specialized device (how it sounds, but it is simply an Arduino).
Here is an example: https://www.ebay.fr/itm/124803897060
You chips should be of type 4164 and 41256.
You can learn more here: https://www.minuszerodegrees.net/5160/ram/5160_ram_flaw_use_of_checkit.htm
There is both the Checkit method and with the debug program. Yours is not 5160, so the memory banks on the motherboard will not be the same.
Actually a memory test program for ELKS will be useful.
